pkg/controller/master/addon/addon.go

```diff
@@ -109,47 +121,191 @@ func (h *Handler) MonitorChanges(key string, aObj *harvesterv1.Addon) (*harveste
+		// logrus.Infof("addon %s status is", aObj.Name, aObj.Status.Status)
+		return aObj, nil
 	}
 
-	hc, err := h.helm.Cache().Get(a.Namespace, a.Name)
+	hc, err := h.helm.Cache().Get(aObj.Namespace, aObj.Name)
 	if err != nil {
-		return a, fmt.Errorf("error querying helmchart in checkHelmChartExists: %v", err)
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("waiting for helmchart %s to be created", aObj.Name)
+			h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+			return aObj, nil
+		}
+
+		return aObj, fmt.Errorf("error querying helmchart %v", err)
 	}
 
 	if hc.Status.JobName == "" {
-		return a, fmt.Errorf("waiting for jobname to be populated in HelmChart status")
+		logrus.Infof("waiting for create jobname to be populated in helmchart status")
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
 	}
 
-	job, err := h.job.Get(a.Namespace, hc.Status.JobName, metav1.GetOptions{})
+	if strings.HasPrefix(hc.Status.JobName, "helm-delete-") {
+		logrus.Infof("the fetched job %s is from last time helmdelete, wait install job", hc.Status.JobName)
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
+	}
+
+	job, err := h.job.Get(aObj.Namespace, hc.Status.JobName, metav1.GetOptions{})
 	if err != nil {
-		return a, fmt.Errorf("error fetching job %s: %v", job.Name, err)
+		logrus.Infof("waiting for job %s to start", job.Name)
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
 	}
 
-	// apply condition //
 	if !isJobComplete(job) {
-		return a, fmt.Errorf("waiting for job %s to complete", job.Name)
+		logrus.Infof("waiting for job %s to complete", job.Name)
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
 	}
 
+	logrus.Infof("addon job %s has finished", job.Name)
+
+	a := aObj.DeepCopy()
+
 	// check Failed before since in jobs with more than 1 container then
 	// even a single failure should count as a failure
 	if job.Status.Failed > 0 {
 		a.Status.Status = harvesterv1.AddonFailed
 	} else if job.Status.Succeeded > 0 {
+		// a.Status.Status = harvesterv1.AddonDeployed
 		a.Status.Status = harvesterv1.AddonDeployed
 	}
 
+	logrus.Infof("addon %s will be set to new status %s", a.Name, a.Status.Status)
+
+	return h.addon.UpdateStatus(a)
+}
+
+// MonitorAddon will track the deployment of Addon
+func (h *Handler) MonitorAddonDisabled(key string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	if aObj == nil || aObj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if aObj.Spec.Enabled {
+		return aObj, nil
+	}
+
+	// has already been successful
+	if aObj.Status.Status == "" {
+		return aObj, nil
+	}
+
+	hc, err := h.helm.Cache().Get(aObj.Namespace, aObj.Name)
+	if err != nil {
+		// disabled
+		if apierrors.IsNotFound(err) {
+			if aObj.Status.Status == "" {
+				logrus.Infof("addon %s: helm chart is gone, addon has already been in init status", aObj.Name)
+				return aObj, nil
+			}
```

Comment:
if `aObj.Status.Status == ""` then this condition will never be let since in L204-205 we already return check for this condition.